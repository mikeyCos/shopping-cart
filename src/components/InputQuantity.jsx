import PropTypes from "prop-types";
import isPressedKeyValid from "../utilities/isPressedKeyValid";
import IncrementIcon from "../assets/icons/increment.svg?react";
import DecrementIcon from "../assets/icons/decrement.svg?react";
import styles from "../styles/InputQuantity.module.css";

const InputQuantity = ({
  quantity,
  setQuantity,
  incrementHandler,
  decrementHandler,
}) => {
  const onChangeHandler = (e) => {
    const value = +e.target.value;
    setQuantity(value);
  };

  const onKeyDownHandler = (e) => {
    const { key } = e;
    switch (true) {
      case /^arrowup$/i.test(key):
        e.preventDefault();
        incrementHandler();
        break;
      case /^arrowdown$/i.test(key):
        e.preventDefault();
        decrementHandler();
        break;
      case isPressedKeyValid(key):
      case !isNaN(key):
        break;
      default:
        e.preventDefault();
    }
  };

  return (
    <div>
      <label htmlFor="quantity"></label>
      <div className={`input-container ${styles["input-btn-container"]}`}>
        <button
          type="button"
          onClick={decrementHandler}
          aria-label="decrement quantity"
          className={styles["input-btn-left"]}
        >
          <DecrementIcon />
        </button>

        <input
          id="quantity"
          name="quantity"
          type="text"
          value={quantity}
          minLength={1}
          maxLength={3}
          autoComplete="off"
          onChange={onChangeHandler}
          onKeyDown={onKeyDownHandler}
          className={`quantity-input ${styles.input}`}
        />
        <button
          type="button"
          onClick={incrementHandler}
          aria-label="increment quantity"
          className={styles["input-btn-right"]}
        >
          <IncrementIcon />
        </button>
      </div>
      <p className={`error-message ${styles["input-error"]}`}></p>
    </div>
  );
};

InputQuantity.propTypes = {
  minValue: PropTypes.number,
  quantity: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  setQuantity: PropTypes.func.isRequired,
  incrementHandler: PropTypes.func.isRequired,
  decrementHandler: PropTypes.func.isRequired,
};

export default InputQuantity;
