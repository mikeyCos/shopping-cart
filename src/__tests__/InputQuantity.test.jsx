import { describe, expect, it, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import InputQuantity from "../components/InputQuantity";
import { useState } from "react";

describe("InputQuantity component", () => {
  it("InputQuantity component is rendered", () => {
    const { container } = render(
      <InputQuantity
        quantity={1}
        setQuantity={() => {}}
        incrementHandler={() => {}}
        decrementHandler={() => {}}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("Quantity input's value is '10'", () => {
    render(
      <InputQuantity
        quantity={10}
        setQuantity={() => {}}
        onChangeHandler={() => {}}
        incrementHandler={() => {}}
        decrementHandler={() => {}}
      />
    );

    const input = screen.getByRole("textbox");
    expect(input.value).toBe("10");
  });

  // Simulates double click to highlight input value then entering a number
  it("Pasting into input fires the setQuantity", async () => {
    const user = userEvent.setup();
    const setQuantity = vi.fn((newQuantity) => {});
    let quantity = 3;

    render(
      <InputQuantity
        quantity={quantity}
        setQuantity={setQuantity}
        incrementHandler={() => {}}
        decrementHandler={() => {}}
      />
    );

    const input = screen.getByRole("textbox");
    await user.dblClick(input);
    await user.paste("4");

    expect(setQuantity).toBeCalledWith("4");
  });

  it("Clicking the incrementButton will increase the quality input's value by 1, initial value is '3", async () => {
    const user = userEvent.setup();

    const ParentComponent = vi.fn(() => {
      // Is it bad practice to define state in unit tests?
      const [quantity, setQuantity] = useState(3);
      return (
        <div>
          <InputQuantity
            quantity={quantity}
            setQuantity={setQuantity}
            incrementHandler={() =>
              setQuantity((prevQuantity) => prevQuantity + 1)
            }
            decrementHandler={() => {}}
          />
        </div>
      );
    });

    render(<ParentComponent />);

    const incrementButton = screen.getByLabelText("increment quantity");
    const input = screen.getByRole("textbox");
    await user.click(incrementButton);

    expect(input.value).toBe("4");
  });
});
