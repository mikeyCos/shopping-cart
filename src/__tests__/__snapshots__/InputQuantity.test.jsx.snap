// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputQuantity component > InputQuantity component is rendered 1`] = `
<div>
  <div>
    <label
      for="quantity"
    />
    <div
      class="input-container _input-btn-container_e589d5"
    >
      <button
        aria-label="decrement quantity"
        class="_input-btn-left_e589d5"
        type="button"
      >
        <svg
          fill="currentColor"
          height="24px"
          viewBox="0 -960 960 960"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M200-440v-80h560v80H200Z"
          />
        </svg>
      </button>
      <input
        autocomplete="off"
        class="quantity-input _input_e589d5"
        id="quantity"
        maxlength="3"
        minlength="1"
        name="quantity"
        type="text"
        value="1"
      />
      <button
        aria-label="increment quantity"
        class="_input-btn-right_e589d5"
        type="button"
      >
        <svg
          fill="currentColor"
          height="24px"
          viewBox="0 -960 960 960"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M440-440H200v-80h240v-240h80v240h240v80H520v240h-80v-240Z"
          />
        </svg>
      </button>
    </div>
    <p
      class="error-message _input-error_e589d5"
    />
  </div>
</div>
`;
